CRIU adds unqualified exports java.base/jdk.crac when CRaC is enabled

JasonFengJ9 opened the pull request:
`CRIU` adds unqualified `exports` `java.base/jdk.crac` when `CRaC` is enabled

Use `Module.implAddExports("jdk.crac")` to export `java.base/jdk.crac` unconditionally when the `CRaC` is enabled;
Refactored `throwExceptionHelper(currentThread, errCode)`;
Also fixed `addPackageDefinition()` usage in `JVM_AddModulePackage()`.

Resolve RTC workitem 150440.

Signed-off-by: Jason Feng <fengj@ca.ibm.com>

tajila commented on runtime/j9vm/java11vmi.c:
```diff
@@ -890,8 +911,25 @@ JVM_DefineModule(JNIEnv * env, jobject module, jboolean isOpen, jstring version,
 							Trc_MODULE_defineModule(currentThread, "java.base", j9mod);
+#if defined(J9VM_OPT_CRAC_SUPPORT)
+							if (J9_ARE_ANY_BITS_SET(vm->checkpointState.flags, J9VM_CRAC_IS_CHECKPOINT_ENABLED)) {
+								UDATA retval = ERRCODE_GENERAL_FAILURE;
```
i think its better to do this at the Java level via `Module::addExports` so if there are changes in the future we dont need to think about this case

JasonFengJ9 replied:
Moved the `addExports` into java.

JasonFengJ9 commented:
> I think you'll need to rebase, we got rid of the static initializer code in InterCRIUSupport, it was causing bootstrap issues with MH

Rebased, moved `Module.implAddExports("jdk.crac")` into `J9VMInternals` to export `java.base/jdk.crac` unconditionally when the `CRaC` is enabled.
@tajila This is ready for another look.

tajila commented on jcl/src/java.base/share/classes/java/lang/J9VMInternals.java:
```diff
@@ -171,6 +171,12 @@ private static void completeInitialization() {
 		}
 		/*[ENDIF] JAVA_SPEC_VERSION >= 9 */
+/*[IF CRAC_SUPPORT]*/
+		if (openj9.internal.criu.InternalCRIUSupport.isCRaCSupportEnabled()) {
```
is this after java.base is created?

JasonFengJ9 replied:
With `-Xtrace:print={module.15} -Xtrace:print=mt,methods={openj9/internal/criu/InternalCRIUSupport.isCRaCSupportEnabled}`, a testcase `org.openj9.criu.TestJDKCRAC` run has following output:
```
20:14:10.705*0x1aa00  module(j9scar).15       - JVM_DefineModule() defines module java.base (0x00007F6AF009A1B8)
20:14:10.752 0x1aa00  module(j9scar).15       - JVM_DefineModule() defines module java.instrument (0x00007F6AF017F478)
.... many other module defines
20:14:10.767 0x1aa00  module(j9scar).15       - JVM_DefineModule() defines module java.xml.crypto (0x00007F6AF01EEC08)
20:14:11.136 0x1aa00              mt.3        > openj9/internal/criu/InternalCRIUSupport.isCRaCSupportEnabled()Z bytecode static method
20:14:11.136 0x1aa00              mt.9        < openj9/internal/criu/InternalCRIUSupport.isCRaCSupportEnabled()Z bytecode static method
```
So `java.base` has been created before `InternalCRIUSupport.isCRaCSupportEnabled()` invocation.
BTW, `TestJDKCRAC` passed as well.

tajila commented:
jenkins test sanity alinux jdk11

tajila commented:
jenkinst test sanity plinux jdk21

JasonFengJ9 commented:
Rebased and resolved the conflict.
Previous PR build https://openj9-jenkins.osuosl.org/job/Test_openjdk11_j9_sanity.functional_aarch64_linux_Personal/266/ - passed

tajila commented on jcl/src/java.base/share/classes/java/lang/J9VMInternals.java:
```diff
@@ -171,6 +171,12 @@ private static void completeInitialization() {
+/*[IF CRAC_SUPPORT]*/
+		if (openj9.internal.criu.InternalCRIUSupport.isCRaCSupportEnabled()) {
+			// export java.base/jdk.crac unconditionally
+			J9VMInternals.class.getModule().implAddExports("jdk.crac"); //$NON-NLS-1$
```
why not Module::addExports ?

JasonFengJ9 replied:
https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/lang/Module.html#addExports(java.lang.String,java.lang.Module)
public [Module](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/lang/Module.html) addExports​([String](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/lang/String.html) pn, [Module](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/lang/Module.html) other)

To export a package unconditionally, the `other` module must be a special [EVERYONE_MODULE](https://github.com/ibmruntimes/openj9-openjdk-jdk11/blob/69eac55903b27058534411d28ab712ea00698bd7/src/java.base/share/classes/java/lang/Module.java#L254) which is a private field within `Module` requiring reflection.
Current invocation avoids the reflection access.

tajila commented:
jenkins test sanity win jdk8

tajila commented:
jenkins test sanity plinux jdk21

JasonFengJ9 commented:
https://openj9-jenkins.osuosl.org/job/Test_openjdk8_j9_sanity.functional_x86-64_windows_Personal_testList_0/402/console
```
11:21:48  init:
11:21:48      [mkdir] Created dir: F:\Users\jenkins\workspace\Test_openjdk8_j9_sanity.functional_x86-64_windows_Personal_testList_0\jvmtest\functional\JLM_Tests
11:21:48      [mkdir] Created dir: F:\Users\jenkins\workspace\Test_openjdk8_j9_sanity.functional_x86-64_windows_Personal_testList_0\aqa-tests\functional\JLM_Tests\bin
11:21:48  
11:21:48  BUILD FAILED
11:21:48  F:\Users\jenkins\workspace\Test_openjdk8_j9_sanity.functional_x86-64_windows_Personal_testList_0\aqa-tests\TKG\scripts\build_test.xml:95: The following error occurred while executing this line:
11:21:48  F:\Users\jenkins\workspace\Test_openjdk8_j9_sanity.functional_x86-64_windows_Personal_testList_0\aqa-tests\functional\JLM_Tests\build.xml:162: The following error occurred while executing this line:
11:21:48  F:\Users\jenkins\workspace\Test_openjdk8_j9_sanity.functional_x86-64_windows_Personal_testList_0\aqa-tests\functional\JLM_Tests\build.xml:54: Execute failed: java.io.IOException: Cannot run program "awk" (in directory "F:\Users\jenkins\workspace\Test_openjdk8_j9_sanity.functional_x86-64_windows_Personal_testList_0\aqa-tests\functional\JLM_Tests"): CreateProcess error=2, The system cannot find the file specified
11:21:48  	at java.lang.ProcessBuilder.start(ProcessBuilder.java:1048)
```
It seems an infra error.

pshipton commented:
It's https://github.com/eclipse-openj9/openj9/issues/18998

keithc-ca commented on runtime/j9vm/java11vmi.c:
```diff
@@ -142,43 +142,55 @@ hashPackageTableDelete(J9VMThread * currentThread, J9ClassLoader * classLoader,
 		case ERRCODE_PACKAGE_WASNT_FOUND:
-			msg = "package was not found";
-			break;
-		default:
```
The `default` case should be retained; using `NLS Failure` instead is not reasonable behavior.

JasonFengJ9 replied:
Opened 
* https://github.com/eclipse-openj9/openj9/pull/19019
